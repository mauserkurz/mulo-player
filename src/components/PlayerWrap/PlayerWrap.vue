<template>
  <v-sheet>
    <v-container class="player-wrap">
      <PlayerHeader @logout="logout"/>

      <TrackList
        :track-list="trackList"
        :current-track="currentTrack"
        @switch-track="switchTrack"
        @get-track="getTrack"
        @cancel-getting-track="cancelGettingTrack"/>

      <footer class="player-wrap__footer">
        <v-card-actions>
          <v-spacer/>
          <v-btn
            color="primary"
            dark
            small
            fab
            @click="switchDrawer(true)">
            <v-icon>mdi-plus</v-icon>
          </v-btn>
        </v-card-actions>

        <Player
          :file="currentTrack.blob"
          :file-name="currentTrack.name"/>
      </footer>
    </v-container>

    <Drawer v-model="isDrawerShown">
      <UploadForm
        :loading="isFileSending"
        :error="sendingFileError"
        @submit="onUploadFormSubmit"/>
    </Drawer>
  </v-sheet>
</template>

<script>
import {
  mapState,
  mapGetters,
  mapActions,
} from 'vuex';
import Player from '@/views/Player/Player.vue';
import PlayerHeader from '@/views/PlayerHeader/PlayerHeader.vue';
import TrackList from '@/views/TrackList/TrackList.vue';
import UploadForm from '@/views/UploadForm/UploadForm.vue';
import Drawer from '@/views/Drawer/Drawer.vue';

export default {
  name: 'PlayerWrap',
  components: {
    Player,
    PlayerHeader,
    TrackList,
    UploadForm,
    Drawer,
  },

  data() {
    return {
      isDrawerShown: false,
    };
  },

  computed: {
    ...mapState('tracks', ['trackList', 'sendingFileError', 'isFileSending']),
    ...mapGetters('tracks', ['currentTrack']),
  },

  methods: {
    ...mapActions('user', ['logout']),
    ...mapActions('tracks', ['getTrack', 'getTrackList', 'switchTrack', 'sendFile', 'cancelGettingTrack']),

    switchDrawer(state) {
      this.isDrawerShown = state;
    },

    async onUploadFormSubmit(file) {
      const isSucceed = await this.sendFile(file);

      if (isSucceed) {
        this.switchDrawer(false);
      }
    },
  },

  async mounted() {
    await this.getTrackList();
  },
};
</script>

<style scoped lang="less">
.player-wrap {
  position: relative;
  z-index: 0;
  width: 543px;
  margin: 20px auto;

  &__footer {
    position: sticky;
    bottom: 20px;
  }
}
</style>
