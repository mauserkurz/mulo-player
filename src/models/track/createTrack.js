const DEFAULT_BLOB = null;

export default ({
  id = '',
  name = '',
  blob = DEFAULT_BLOB,
  lastModifiedDate = new Date().getTime(),
  isLoading = false,
}) => ({
  id: `${id}`,
  name,
  blob,
  lastModifiedDate,
  isLoading,

  get isLoaded() {
    return this.blob !== DEFAULT_BLOB;
  },
});
