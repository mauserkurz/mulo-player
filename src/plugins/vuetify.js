import Vue from 'vue';
import Vuetify from 'vuetify';
import { THEME as theme } from '@/constants';

Vue.use(Vuetify);

export default new Vuetify({ theme });
