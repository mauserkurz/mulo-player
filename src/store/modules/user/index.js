import { path, pathOr } from 'ramda';
import utils from '@/utils';
import api from '@/api/';
import router from '@/router';
import { AUTH_FORM_TYPE_MAP, STATUS_MAP, API_ERROR_MAP } from '@/const';

const getErrorKey = response => pathOr([], ['data', 'errors', 0, 'message'], response);

export default {
  namespaced: true,

  state: {
    userID: '',
    login: '',
    error: '',
    isAuthLoading: false,
    isUserDataLoading: false,
    isLogoutProcessing: false,
  },

  getters: {
    isAuthorized(state) {
      return !!state.userID;
    },
  },

  mutations: {
    SET_USER_ID(state, id) {
      state.userID = id;
    },

    SET_LOGIN(state, login) {
      state.login = login;
    },

    SET_AUTH_ERROR(state, error) {
      state.error = error;
    },

    SET_IS_AUTH_LOADING(state, mode) {
      state.isAuthLoading = mode;
    },

    SET_IS_USER_DATE_LOADING(state, mode) {
      state.isUserDataLoading = mode;
    },

    SET_IS_LOGOUT_PROCESSING(state, mode) {
      state.isLogoutProcessing = mode;
    },
  },

  actions: {
    setUser({ commit }, data) {
      commit('SET_USER_ID', data.user_id);
      commit('SET_LOGIN', data.login);
      router.push({ path: 'player' });
    },

    async auth({ commit, dispatch }, { login, password, type }) {
      const methodMap = {
        [AUTH_FORM_TYPE_MAP.SIGN_IN]: 'signIn',
        [AUTH_FORM_TYPE_MAP.SIGN_UP]: 'signUp',
      };
      let response;

      commit('SET_IS_AUTH_LOADING', true);
      try {
        const request = api[methodMap[type]];

        response = await request({ login, password });
      } catch (error) {
        const errorKey = getErrorKey(error.response);
        const message = API_ERROR_MAP.AUTH[errorKey];

        // For 401 status show message from map
        // otherwise show message generated from error
        if (path(['response', 'status'], error) === STATUS_MAP.UNAUTHORIZED && message) {
          commit('SET_AUTH_ERROR', message);
        } else {
          commit('SET_AUTH_ERROR', utils.createMessage(error));
        }
        return;
      } finally {
        commit('SET_IS_AUTH_LOADING', false);
      }
      // After successful request clear error, write user id and login
      // then show player itself
      commit('SET_AUTH_ERROR', '');
      dispatch('setUser', response.data);
    },

    async getUser({ commit, dispatch }) {
      let response;

      commit('SET_IS_USER_DATE_LOADING', true);
      try {
        response = await api.getUser();
      } catch (error) {
        if (path(['response', 'status'], error) !== STATUS_MAP.UNAUTHORIZED) {
          window.console.error(error);
        }
        return;
      } finally {
        commit('SET_IS_USER_DATE_LOADING', false);
      }
      dispatch('setUser', response.data);
    },

    async logout({ state, commit }) {
      if (state.isLogoutProcessing) {
        return;
      }
      commit('SET_IS_LOGOUT_PROCESSING', true);

      try {
        await api.logout();
      } catch (error) {
        if (path(['response', 'status'], error) !== STATUS_MAP.UNAUTHORIZED) {
          window.console.error(error);
        }
        commit('SET_IS_LOGOUT_PROCESSING', false);
        return;
      }

      commit('SET_USER_ID', '');
      commit('SET_LOGIN', '');
      await router.push({ path: '/' });
      commit('SET_IS_LOGOUT_PROCESSING', false);
    },

    clearError({ commit }) {
      commit('SET_AUTH_ERROR', '');
    },
  },
};
