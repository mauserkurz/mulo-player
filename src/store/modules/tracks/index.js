import axios from 'axios';
import { API_ERROR_MAP } from '@/constants';
import utils from '@/utils';
import api from '@/api/';
import createTrack from '@/models/track/createTrack';

const findTrack = state => trackID => state.trackList.find(({ id }) => id === trackID);

export default {
  namespaced: true,

  state: {
    currentTrackID: null,
    trackList: [],
    sendingFileError: '',
    isFileSending: false,
    loadTrackCancelSourceMap: {},
  },

  getters: {
    currentTrack(state, getters) {
      if (state.currentTrackID === null) {
        return {};
      }
      return getters.track(state.currentTrackID);
    },

    loadTrackCancelSourceMap: state => trackID => state.loadTrackCancelSourceMap[trackID],

    track: findTrack,
  },

  mutations: {
    SET_TRACK_LIST(state, list) {
      state.trackList = list;
    },

    SET_CURRENT_TRACK_ID(state, trackID) {
      state.currentTrackID = trackID;
    },

    SET_AUDIO_BLOB(state, { trackID, blob }) {
      findTrack(state)(trackID).blob = blob;
    },

    SET_TRACK_LOADING_STATE(state, { trackID, isLoading }) {
      findTrack(state)(trackID).isLoading = isLoading;
    },

    ADD_TRACK_TO_LIST(state, { id, name, blob }) {
      state.trackList.push(createTrack({ id, name, blob }));
    },

    SET_SENDING_FILE_ERROR(state, error) {
      state.sendingFileError = error;
    },

    SET_IS_FILE_SENDING(state, mode) {
      state.isFileSending = mode;
    },

    ADD_LOAD_TRACK_CANCEL_SOURCE(state, { trackID, source }) {
      state.loadTrackCancelSourceMap[trackID] = source;
    },

    DELETE_LOAD_TRACK_CANCEL_SOURCE(state, trackID) {
      delete state.loadTrackCancelSourceMap[trackID];
    },
  },

  actions: {
    async getTrackList({ rootState, commit, dispatch }) {
      const { userID } = rootState.user;
      let response = {};

      try {
        response = await api.getAllTracksInfo({ userID });
      } catch (error) {
        window.console.error(error);
        return;
      }
      const { tracks } = response.data;

      if (tracks.length > 0) {
        commit('SET_TRACK_LIST', tracks.map(createTrack));
        await dispatch('switchTrack', response.data.tracks[0].id);
      } else {
        commit('SET_TRACK_LIST', []);
        commit('SET_CURRENT_TRACK_ID', null);
      }
    },

    async getTrack({ rootState, getters, commit }, trackID) {
      const { userID } = rootState.user;
      const source = axios.CancelToken.source();

      if (getters.track(trackID).blob) {
        return;
      }
      let response = {};

      commit('SET_TRACK_LOADING_STATE', { trackID, isLoading: true });
      commit('ADD_LOAD_TRACK_CANCEL_SOURCE', { trackID, source });
      try {
        response = await api.loadTrack({ userID, trackID, cancelToken: source.token });
      } catch (error) {
        if (error.message !== API_ERROR_MAP.CANCELED_BY_USER) {
          window.console.error(error);
        }
        return;
      } finally {
        commit('DELETE_LOAD_TRACK_CANCEL_SOURCE', trackID);
        commit('SET_TRACK_LOADING_STATE', { trackID, isLoading: false });
      }
      commit('SET_AUDIO_BLOB', { trackID, blob: response.data });
    },

    cancelGettingTrack({ getters, commit }, trackID) {
      const source = getters.loadTrackCancelSourceMap(trackID);

      source.cancel(API_ERROR_MAP.CANCELED_BY_USER);
      commit('SET_TRACK_LOADING_STATE', { trackID, isLoading: false });
    },

    async switchTrack({ commit, dispatch }, ID) {
      await dispatch('getTrack', ID);
      commit('SET_CURRENT_TRACK_ID', ID);
    },

    async sendFile({ rootState, commit }, file) {
      const { userID } = rootState.user;
      let response;

      commit('SET_IS_FILE_SENDING', true);
      try {
        response = await api.sendTrack({ userID, file });
      } catch (error) {
        commit('SET_SENDING_FILE_ERROR', utils.createMessage(error));
        return false;
      } finally {
        commit('SET_IS_FILE_SENDING', false);
      }
      const { id, name } = response.data.tracks[0];

      commit('ADD_TRACK_TO_LIST', { id, name, blob: file });
      commit('SET_SENDING_FILE_ERROR', '');
      return true;
    },
  },
};
