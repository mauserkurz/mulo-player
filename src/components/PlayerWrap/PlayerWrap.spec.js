import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import flushPromises from 'flush-promises';
import clone from 'ramda/src/clone';
import Vuex from 'vuex';
import vuetify from '@/plugins/vuetify';
import user from '@/store/modules/user/';
import tracks from '@/store/modules/tracks/';
import createTrack from '@/models/track/createTrack';
import PlayerWrap from './PlayerWrap.vue';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Component PlayerWrap', () => {
  const createWrapper = ({ isWithoutStubs = false, modules = {}, ...options } = {}) => {
    const renderer = isWithoutStubs ? mount : shallowMount;
    tracks.actions.getTrackList = () => Promise.resolve();
    const store = new Vuex.Store({
      namespaced: true,

      modules: Object.assign({ user, tracks }, modules),
    });
    const wrapper = renderer(PlayerWrap, {
      localVue,
      vuetify,
      store,
      ...options,
    });

    return { wrapper };
  };

  describe('snapshots', () => {
    it('should match snapshot', () => {
      const tracksCopy = clone(tracks);

      tracksCopy.state.trackList = [
        createTrack({
          id: 0,
          name: 'Band 0 - Song 0',
          isLoading: true,
        }),
        createTrack({
          id: 1,
          name: 'Band 1 - Song 1',
          blob: new Blob([], { type: 'audio/mpeg' }),
        }),
        createTrack({
          id: 2,
          name: 'Band 2 - Song 2',
        }),
      ];
      tracksCopy.actions.getTrackList = () => Promise.resolve();
      const { wrapper } = createWrapper({ modules: { tracks: tracksCopy }, isWithoutStubs: true });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('events', () => {
    it('should call logout action after click on logout button', () => {
      const userCopy = clone(user);
      const spy = jest.fn();

      userCopy.actions.logout = spy;
      const { wrapper } = createWrapper({ modules: { user: userCopy }, isWithoutStubs: true });

      wrapper.find('.player-header').vm.$emit('logout');
      expect(spy).toHaveBeenCalled();
    });

    it('should open drawer after click on cross button', async () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      wrapper.find('.player-wrap__footer .v-btn').trigger('click');
      await wrapper.vm.$nextTick();
      expect(wrapper.find('.drawer').vm.value).toBe(true);
    });

    it('should call clearError action on clear-form event from UploadForm', () => {
      const tracksCopy = clone(tracks);
      const spy = jest.fn();

      tracksCopy.actions.clearError = spy;
      tracksCopy.actions.getTrackList = () => Promise.resolve();
      const { wrapper } = createWrapper({ modules: { tracks: tracksCopy }, isWithoutStubs: true });

      wrapper.find('.upload-form').vm.$emit('clear-form');
      expect(spy).toHaveBeenCalled();
    });

    it('should call sendFile action on submit event from UploadForm', () => {
      const file = new File([], 'track', { type: 'audio/mpeg' });
      const tracksCopy = clone(tracks);
      const spy = jest.fn();

      tracksCopy.actions.sendFile = spy;
      tracksCopy.actions.getTrackList = () => Promise.resolve();
      const { wrapper } = createWrapper({ modules: { tracks: tracksCopy }, isWithoutStubs: true });

      wrapper.find('.upload-form').vm.$emit('submit', file);
      expect(spy).toHaveBeenCalled();
    });

    it('should open drawer after successful file sending', async () => {
      const file = new File([], 'track', { type: 'audio/mpeg' });
      const tracksCopy = clone(tracks);

      tracksCopy.actions.sendFile = () => Promise.resolve(true);
      tracksCopy.actions.getTrackList = () => Promise.resolve();
      const { wrapper } = createWrapper({ modules: { tracks: tracksCopy }, isWithoutStubs: true });

      wrapper.find('.upload-form').vm.$emit('submit', file);
      await flushPromises();
      expect(wrapper.find('.drawer').vm.value).toBe(false);
    });

    it('should call cancelGettingTrack action on cancel-getting-track event from TrackList', () => {
      const trackID = 5;
      const tracksCopy = clone(tracks);
      const spy = jest.fn();

      tracksCopy.actions.cancelGettingTrack = spy;
      tracksCopy.actions.getTrackList = () => Promise.resolve();
      const { wrapper } = createWrapper({ modules: { tracks: tracksCopy }, isWithoutStubs: true });

      wrapper.find('.track-list').vm.$emit('cancel-getting-track', trackID);
      expect(spy).toHaveBeenCalled();
    });

    it('should call updateTrackList action on update-track-list event from TrackList', () => {
      const tracksCopy = clone(tracks);
      const spy = jest.fn();
      const trackList = [];

      tracksCopy.actions.updateTrackList = spy;
      tracksCopy.actions.getTrackList = () => Promise.resolve();
      const { wrapper } = createWrapper({ modules: { tracks: tracksCopy }, isWithoutStubs: true });

      wrapper.find('.track-list').vm.$emit('update-track-list', trackList);
      expect(spy.mock.calls[0][1]).toEqual(trackList);
    });
  });
});
