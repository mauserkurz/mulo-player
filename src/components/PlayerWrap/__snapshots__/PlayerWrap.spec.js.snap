// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component PlayerWrap snapshots should match snapshot 1`] = `
<div class="v-sheet theme--light">
  <div class="container player-wrap">
    <div class="row player-header align-center justify-space-between"><a href="/player" title="Music lover"><img alt="App logo" src="@/assets/logo.svg" width="64" height="64"></a> <button type="button" class="v-btn v-btn--depressed v-btn--flat v-btn--outlined theme--light v-size--default primary--text"><span class="v-btn__content">
    Logout
  </span></button></div>
    <div role="list" class="v-list track-list v-sheet v-sheet--tile theme--light v-list--rounded">
      <div role="listbox" class="v-item-group theme--light v-list-item-group primary--text">
        <div class="row track-list-setting">
          <div class="col col-4">
            <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
              <div class="v-input__control">
                <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-12" class="v-input__slot">
                  <div class="v-select__slot">
                    <div class="v-select__selections">
                      <div class="v-select__selection v-select__selection--comma">Search name</div><input id="input-12" readonly="readonly" type="text" aria-readonly="true" autocomplete="off">
                    </div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                    </div><input type="hidden" value="text">
                  </div>
                  <div class="v-menu">
                    <div class="v-menu__content theme--light " style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                </div>
              </div>
            </div>
          </div>
          <div class="col col-4">
            <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
              <div class="v-input__control">
                <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-19" class="v-input__slot">
                  <div class="v-select__slot">
                    <div class="v-select__selections">
                      <div class="v-select__selection v-select__selection--comma">From A to Z</div><input id="input-19" readonly="readonly" type="text" aria-readonly="true" autocomplete="off">
                    </div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                    </div><input type="hidden" value="abc">
                  </div>
                  <div class="v-menu">
                    <div class="v-menu__content theme--light " style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="v-menu track-list-search">
          <div class="v-input theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--solo-inverted v-text-field--solo-flat v-text-field--enclosed">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-input__prepend-inner">
                  <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">search</i></div>
                </div>
                <div class="v-text-field__slot"><label for="input-28" class="v-label theme--light" style="left: 0px; position: absolute;">Search</label><input id="input-28" type="text"></div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link material-icons theme--light"></i></div>
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
              </div>
            </div>
          </div>
          <div role="menu" class="v-menu__content theme--light " style="max-height: auto; min-width: 290px; max-width: 290px; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
        </div>
        <div><span name="slide-y-transition"><div tabindex="0" transition="slide-y-transition" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light track-list-item"><div class="v-list-item__icon"><div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="v-progress-circular v-progress-circular--indeterminate" style="height: 24px; width: 24px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="24 24 48 48" style="transform: rotate(0deg);"><circle fill="transparent" cx="48" cy="48" r="20" stroke-width="8" stroke-dasharray="125.664" stroke-dashoffset="125.66370614359172px" class="v-progress-circular__overlay"></circle></svg><div class="v-progress-circular__info"></div></div></div> <div class="v-list-item__content"><div class="v-list-item__title">Band 0 - Song 0</div></div></div><div tabindex="0" transition="slide-y-transition" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light track-list-item"><div class="v-list-item__icon"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link mdi mdi-check-circle theme--light"></i></div> <div class="v-list-item__content"><div class="v-list-item__title">Band 1 - Song 1</div></div></div><div tabindex="0" transition="slide-y-transition" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light track-list-item"><div class="v-list-item__icon"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link mdi mdi-cloud-download theme--light"></i></div> <div class="v-list-item__content"><div class="v-list-item__title">Band 2 - Song 2</div></div></div></span></div>
      </div>
    </div>
    <footer class="player-wrap__footer">
      <div class="v-card__actions">
        <div class="spacer"></div> <button type="button" class="v-btn v-btn--contained v-btn--fab v-btn--round theme--dark v-size--small primary"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-plus theme--dark"></i></span></button>
      </div>
      <div class="player"><button title="Stop" type="button" class="control"><img src="" alt="stop icon" width="20" height="20">
          <!----> </button> <button title="Play/Pause" type="button" class="control"><img src="" alt="pause icon" width="20" height="20"> <img src="" alt="play icon" width="20" height="20" style="display: none;"> </button>
        <div class="time-line">
          <div title="Time played : Total time" class="time-line__progress">
            <div class="time-line__line"></div>
          </div>
          <div class="time-line__timers">
            <div class="time-line__current">00:00</div>
            <div class="time-line__total">00:00</div>
          </div>
        </div> <button title="Download" type="button" class="control"><img src="" alt="download icon" width="20" height="20">
          <!----> </button> <button title="Loop/Once" type="button" class="control"><img src="" alt="once icon" width="20" height="20"> <img src="" alt="loop icon" width="20" height="20" style="display: none;"> </button> <button title="Mute" type="button" class="control"><img src="" alt="sound icon" width="20" height="20"> <img src="" alt="mute icon" width="20" height="20" style="display: none;"> </button> <button title="Volume" type="button" class="control"><img src="" alt="volume icon" width="20" height="20">
          <!----> <input type="range" min="0" step="0.01" max="1" style="display: none;"></button>
      </div>
    </footer>
  </div>
  <aside class="drawer v-navigation-drawer v-navigation-drawer--close v-navigation-drawer--fixed v-navigation-drawer--is-mobile v-navigation-drawer--right v-navigation-drawer--temporary theme--light" style="height: 100%; top: 0px; transform: translateX(100%); width: 256px;">
    <div class="v-navigation-drawer__content">
      <div tabindex="-1" class="v-list-item theme--light">
        <div class="v-list-item__content">
          <div class="v-list-item__title title"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small primary--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-close-circle-outline theme--light"></i></span></button>
            Upload music
          </div>
        </div>
      </div>
      <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
      <div class="container">
        <div class="row align-center justify-center">
          <div class="col col-12">
            <div class="upload-form v-card v-sheet theme--light">
              <form novalidate="novalidate" class="v-form">
                <div class="v-card__text">
                  <div class="v-input theme--light v-text-field v-text-field--placeholder v-file-input">
                    <div class="v-input__prepend-outer">
                      <div class="v-input__icon v-input__icon--prepend"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link mdi mdi-file-music theme--light"></i></div>
                    </div>
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <div class="v-text-field__slot"><label for="input-65" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Music file</label>
                          <div class="v-file-input__text v-file-input__text--placeholder">Select mp3 file</div><input accept="audio/mp3, audio/mpeg" id="input-65" placeholder="Select mp3 file" type="file">
                        </div>
                        <div class="v-input__append-inner">
                          <div class="v-input__icon v-input__icon--"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link material-icons theme--light"></i></div>
                        </div>
                      </div>
                      <div class="v-text-field__details">
                        <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="v-card__text" style="display: none;">
                  <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense error">
                    <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark"></i>
                      <div class="v-alert__content">

                      </div>
                    </div>
                  </div>
                </div>
                <div class="v-card__actions">
                  <div class="spacer"></div> <button type="submit" class="v-btn v-btn--contained theme--light v-size--default primary"><span class="v-btn__content">
        Save
      </span></button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-navigation-drawer__border"></div>
  </aside>
</div>
`;
